### ISSUE TITLE
Subscribe to Blazor Web Assembly AuthenticationStateChange Event

### ISSUE THREAD

**Issue by nssidhu** [dotnet/aspnetcore#40163]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I want to execute some code in the .razor.cs page whenever user logs in or logs out.

When i implemented this code, the AuhtenticationStateChanged event keeps on firing for infinite time. 
Also i could not find any sample code which describes how to check the current state in this event, i.e Login or Logout
	```
protected override async Task OnInitializedAsync()
	{
		_CustomAuthProvider.AuthenticationStateChanged += AuhtenticationStateChanged;
      }

    private async void AuhtenticationStateChanged(Task<AuthenticationState> task)
	{
		//try
		//{
		//	var user = (await task).User;

		//	if (user.Identity.IsAuthenticated)
		//	{
		//		// Perform an action only available to authenticated (signed-in) users.
		//		IsLoggedIn = true;
		//	}
		//	else
		//	{
		//		IsLoggedIn = false;
		//	}
		//}
		//catch(Exception ex)
		//{
		//	Console.WriteLine("Error .." + ex.Message);
		//}
		
	}
```

### Expected Behavior

Fire event only once when change of Authentication state occurs like user successfully logged in or user successfully logged out.

### Steps To Reproduce

1) Create CustomAuthenticationStateProvider
2) Register it for DI in program.cs
3) in any page like NavMenu.razor.cs write the following code.

  protected override async Task OnInitializedAsync()
	{
		_CustomAuthProvider.AuthenticationStateChanged += AuhtenticationStateChanged;
      }

private async void AuhtenticationStateChanged(Task<AuthenticationState> task)
	{
		//try
		//{
		//	var user = (await task).User;

		//	if (user.Identity.IsAuthenticated)
		//	{
		//		// Perform an action only available to authenticated (signed-in) users.
		//		IsLoggedIn = true;
		//	}
		//	else
		//	{
		//		IsLoggedIn = false;
		//	}
		//}
		//catch(Exception ex)
		//{
		//	Console.WriteLine("Error .." + ex.Message);
		//}
		
	}

Place a debug point AuhtenticationStateChanged the event, you will see that this event keeps on firing for infinite times.

### Exceptions (if any)

Every time event fires, i get casting errors.

### .NET Version

6.0.200-preview.22055.15

**Comment by javiercn** [comment]
@nssidhu thanks for contacting us.

You can't replace the authenticationstate provider in webassembly without replacing other components that are associated with it.

**Comment by nssidhu** [comment]
@javiercn I did not understand your suggestion, can you elaborate on it ?

I followed https://docs.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-6.0#implement-a-custom-authenticationstateprovider. Created necessary class registered them in program.cs update main.razor with <CascadingAuthenticationState>

Custom Authentication is working fine for me, no issues there. 

My question is specific to the subscribing to the _CustomAuthProvider.AuthenticationStateChanged

**Comment by stephajn** [comment]
I am experiencing this problem with the built in AuthenticationStateProvider in Blazor WASM.  It seems like the entire flow of signing a user in causes all of these moving pieces to fire an unpredictable number of times over and over and over again even when nothing has changed.

There is a serious problem with the framework here as has been reported by numerous other users on different issues, all with the same response.  "Moved to .NET 7 Milestone".  This is broken **now** in .NET 6, which is supposed to be a LTS release. This needs to be addressed now, not later.

**Comment by SteveSandersonMS** [comment]
@stephajn It's difficult for us to take action on a generic concern. If you're able to file a new issue giving specific repro steps for a problem you're facing, we'd be happy to take a look. Hope that's OK.

**Comment by stephajn** [comment]
Thanks, Steve. I will create a repository that shows what is happening and open a separate issue for it.
